lib/Dialect/Torch/IR/TorchOps.cpp

```diff
@@ -715,16 +715,58 @@ OpFoldResult AtenNeBoolOp::fold(FoldAdaptor adaptor) {
   return IntegerAttr::get(IntegerType::get(getContext(), 1), a != b);
 }
 
+//===----------------------------------------------------------------------===//
+// AtenUnsqueezeOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult AtenUnsqueezeOp::fold(FoldAdaptor adaptor) {
+  auto ty = dyn_cast<BaseTensorType>(getSelf().getType());
+  auto rty = dyn_cast<BaseTensorType>(getType());
+  if (!rty.hasDtype())
+    return {};
+
+  if (auto attr = dyn_cast_or_null<DenseElementsAttr>(adaptor.getSelf())) {
+    auto aty = dyn_cast<RankedTensorType>(attr.getType());
+    if (rty.hasSizes() && rty.areAllSizesKnown() && attr.isSplat()) {
+      auto naty = RankedTensorType::get(rty.getSizes(), aty.getElementType());
+      return DenseElementsAttr::get(naty, attr.getSplatValue<Attribute>());
+    }
+  }
+
+  if (getSelf().getType() != getResult().getType())
+    return nullptr;
+  if (ty && rty) {
+    if (ty.hasSizes() && rty.hasSizes() &&
+        ty.getSizes().size() == rty.getSizes().size())
+      return getSelf();
+  }
+  return nullptr;
+}
+
```

Comment:
I haven't written a folder before, so I have a few questions:

1. What is the difference between "getSelf()" and "adaptor.getSelf()" I assume "getSelf()" is getting "self"/the input arg from the Unsqueeze op being folded, but I'm not sure about the latter. 
2. I'm similarly confused about getType(). Is this equivalent to "op.getResult().getType()"? 
3. Is the purpose of this folder to convert the UnsqueezeOp to a splat array if possible and to completely ignore the unsqueeze if no dims were actually unsqueezed?

Reply:
1. Type conversion - `getSelf()` gets the non-converted version. I.e. the source version. While `adaptor.getSelf()` gets the version to the target operation. In this case it gives either the 'torch' types version vs the `mlir` version of the tensor.

2. `getType()` vs `op.getResult().getType()` is no difference. Most of this is likely force of habit from before `getType()` worked standalone.

3. Thats about it. If something is a splat we can just proxy it through, if the rank doesn't change we can remove the operation.